src/gwt/src/org/rstudio/studio/client/application/ui/CommandPalette.java

```diff
@@ -0,0 +1,442 @@
+         commandList_.add(entry);
+      }
+
+      
+      // Handle most keystrokes on KeyUp so that the contents of the text box
+      // have already been changed
+      searchBox_.addKeyUpHandler((evt) ->
+      {
+         if (evt.getNativeKeyCode() == KeyCode.ESC)
+         {
+            // Pressing ESC dismisses the host (removing the palette popup)
+            host_.dismiss();
+         }
+         else if (evt.getNativeKeyCode() == KeyCode.ENTER)
+         {
+            // Enter runs the selected command
+            invokeSelection();
+         }
+         else
+         {
+            // Just update the filter if the text has changed
+            String searchText = searchBox_.getText();
+            if (!StringUtil.equals(searchText_, searchText))
+            {
+               searchText_ = searchText;
+               applyFilter();
+            }
+         }
+      });
+
+      // Up and Down arrows need to be handled on KeyDown to account for
+      // repetition (a held arrow key will generate multiple KeyDown events and
+      // then a single KeyUp when released)
+      searchBox_.addKeyDownHandler((evt) -> 
+      {
+         // Ignore the Tab key so we don't lose focus accidentally (there is
+         // only one focusable element in the palette and we don't want Tab to
+         // dismiss it)
+         if (evt.getNativeKeyCode() == KeyCode.TAB)
+         {
+            evt.stopPropagation();
+            evt.preventDefault();
+            return;
+         }
+
+         // Ignore modified arrows so that e.g. Shift Up/Down to select the
+         // contents of the textbox work as expected
+         if (evt.isAnyModifierKeyDown())
+            return;
+         
+         if (evt.getNativeKeyCode() == KeyCode.UP)
+         {
+            // Directional keys often trigger behavior in textboxes (e.g. moving
+            // the cursor to the beginning/end of text) but we're hijacking them
+            // to do navigation in the results list, so disable that.
+            evt.stopPropagation();
+            evt.preventDefault();
+            moveSelection(-1);
+         }
+         else if (evt.getNativeKeyCode() == KeyCode.DOWN)
+         {
+            evt.stopPropagation();
+            evt.preventDefault();
+            moveSelection(1);
+         }
+      });
+   }
+   
+   /**
+    * Filter the commands by the current contents of the search box
+    */
+   private void applyFilter()
+   {
+      int matches = 0;
+
+      // Split the search text into a series of lowercase words. This provides a
+      // kind of partial fuzzy matching, so that e.g., "new py" matches the command
+      // "Create a new Python script".
+      String[] needles = searchBox_.getText().toLowerCase().split("\\s+");
+      
+      for (CommandPaletteEntry entry: entries_)
+      {
+         String hay = entry.getLabel().toLowerCase();
+         boolean matched = true;
+         for (int i = 0; i < needles.length; i++)
```

Comment:
IntelliJ nit: could use `for (String needle : needles)`

Reply:
Updated.